:sparkles: binding controller code improvements

vMaroon opened the pull request:
## Summary
This PR:
- Implements left-over comments from #1831's review:
  - This involves documentation refactoring and general code clarity improvements
  - The GVKToGVRMapper is completely removed
    - Controller::handleObject receives object resource name alongside the object
    - This required changing informer maps (and the rest of the family) to be keyed by GVR and not GVK
- Updates BindingPolicyResolver::GenerateBinding's implementation to generate bindings with sorted workload objects and destination arrays
- Fixes Controller::getObjectFromKey to not have an unexpected return path

## Related issue(s)

Fixes #1790
Fixes #1716
Fixes #1845

vMaroon commented:
/cc @MikeSpreitzer 
/cc @nirrozenbaum 
/cc @waltforme

MikeSpreitzer commented on pkg/binding/selectors.go:
```diff
@@ -93,8 +96,8 @@ func (c *Controller) updateResolutions(ctx context.Context, objIdentifier util.O
-				utilruntime.HandleError(fmt.Errorf("failed to note object (identifier: %v) - %w",
-					objIdentifier, err))
+				logger.V(4).Info("skipped EnsureObjectIdentifierWithVersion for object-identifier (%v) and "+
+					"bindingpolicy %s because the latter was deleted", objIdentifier, bindingPolicy.GetName())
```
This is a call on `Info`, so you need logging keys as well as values here.

MikeSpreitzer commented on pkg/binding/selectors.go:
```diff
@@ -93,8 +96,8 @@ func (c *Controller) updateResolutions(ctx context.Context, objIdentifier util.O
 				// this case can occur if a bindingpolicy resolution was deleted AFTER
 				// starting this iteration and BEFORE getting to the NoteObject function,
 				// which occurs if a bindingpolicy was deleted in this time-window.
```
FYI: Now the window is even smaller, it opens after the new check above for `c.bindingPolicyResolver.ResolutionExists`.

vMaroon commented:
@MikeSpreitzer addressed all comments, and added the above commit to stop repeatedly starting informers on API resources. 
The latter was observed during a test, where every update to a binding-policy would start informers on Binding and BindingPolicy APIs.

MikeSpreitzer commented on pkg/binding/bindingpolicy-resolution.go:
```diff
@@ -215,14 +215,44 @@ func bindingObjectRefSetFromBindingWorkload(bindingSpecWorkload *v1alpha1.Downsy
+	sortedDestiantions := make([]v1alpha1.Destination, 0, len(destinationsStringSet))
+
+	// sort destinations
+	sortedDestinationStrings := destinationsStringSet.UnsortedList()
+	slices.Sort(sortedDestinationStrings)
```
FYI:

```suggestion
	sortedDestinationStrings := destinationsStringSet.List()
```

MikeSpreitzer replied:
Please let me mark my own conversations as "resolved".

vMaroon replied:
@MikeSpreitzer apologies, it was marked resolved automatically when I committed your suggestion through the UI here.

MikeSpreitzer replied:
Oops, sorry, my suggestion was buggy. `List` is a func in the `sets` package, not a method of `sets.Set`.

vMaroon replied:
Was wondering how I missed that method. So reverting is sufficient?

MikeSpreitzer replied:
Reverting would work. So would changing the code to the following.

```go
	sortedDestinationStrings := sets.List(destinationsStringSet)
```

vMaroon replied:
Done.

vMaroon commented:
@MikeSpreitzer I've reverted the sorting change

vMaroon commented on pkg/binding/bindingpolicy-resolution.go:
```diff
@@ -215,14 +215,43 @@ func bindingObjectRefSetFromBindingWorkload(bindingSpecWorkload *v1alpha1.Downsy
+func destinationsStringSetToSortedDestinations(destinationsStringSet sets.Set[string]) []v1alpha1.Destination {
+	sortedDestiantions := make([]v1alpha1.Destination, 0, len(destinationsStringSet))
+
+	// sort destinations
```
Done.

MikeSpreitzer reviewed (approved):
/lgtm
/approve
Thanks!
